daemon/logger/ring.go

```diff
@@ -0,0 +1,219 @@
+	}
+	l.run()
+	return l
+}
+
+// NewRingLogger creates a new Logger that is implemented as a RingBuffer wrapping
+// the passed in logger.
+func NewRingLogger(driver Logger, logCtx Info, maxSize int64) Logger {
+	if maxSize < 0 {
+		maxSize = defaultRingMaxSize
+	}
+	l := newRingLogger(driver, logCtx, maxSize)
+	if _, ok := driver.(LogReader); ok {
+		return &ringWithReader{l}
+	}
+	return l
+}
+
+// Log queues messages into the ring buffer
+func (r *RingLogger) Log(msg *Message) error {
+	select {
+	case <-r.ctx.Done():
+		return r.ctx.Err()
+	default:
+	}
+	return r.buffer.Enqueue(msg)
+}
+
+// Name returns the name of the underlying logger
+func (r *RingLogger) Name() string {
+	return r.l.Name()
+}
+
+// Close closes the logger
+func (r *RingLogger) Close() error {
+	r.cancel()
+	r.buffer.Close()
+	// empty out the queue
+	for _, msg := range r.buffer.Drain() {
+		if err := r.l.Log(msg); err != nil {
+			logrus.WithField("driver", r.l.Name()).WithField("container", r.logCtx.ContainerID).Errorf("Error writing log message: %v", r.l)
+			break
+		}
+	}
+	return r.l.Close()
+}
+
+// run consumes messages from the ring buffer and forwards them to the underling
+// logger.
+// This is run in a goroutine when the RingLogger is created
+func (r *RingLogger) run() {
+	go func() {
+		for {
+			select {
+			case <-r.ctx.Done():
+				return
+			default:
+				msg, err := r.buffer.Dequeue()
+				if err != nil {
+					// buffer is closed
+					return
+				}
+				if err := r.l.Log(msg); err != nil {
+					logrus.WithField("driver", r.l.Name()).WithField("container", r.logCtx.ContainerID).Errorf("Error writing log message: %v", r.l)
+				}
+			}
+		}
+	}()
+}
+
+type messageRing struct {
+	mu sync.Mutex
+	// singals callers of `Dequeue` to wake up either on `Close` or when a new `Message` is added
+	wait *sync.Cond
+
+	sizeBytes int64 // current buffer size
+	cap       int64 // max buffer size size
+	queue     []*Message
+	closeErr  error // error set on close
+}
+
+func newRing(maxBytes int64) *messageRing {
+	queueSize := 1000
+	if maxBytes == 0 || maxBytes == 1 {
+		// With 0 or 1 max byte size, the maximum size of the queue would only ever be 1
+		// message long.
+		queueSize = 1
+	}
+
+	r := &messageRing{queue: make([]*Message, 0, queueSize), cap: maxBytes}
+	r.wait = sync.NewCond(&r.mu)
+	return r
+}
+
+// Enqueue adds a message to the buffer queue
+// If the message is too big for the buffer it drops the oldest messages to make room
+// If there are no messages in the queue and the message is still too big, it adds the message anyway.
+func (r *messageRing) Enqueue(m *Message) error {
+	m.Size = len(m.Line) // ensure size is set
+	mSize := int64(m.Size)
+
+	r.mu.Lock()
+	for {
+		if r.closeErr != nil {
+			r.mu.Unlock()
+			return r.closeErr
+		}
+
+		if len(r.queue) == 0 || mSize+r.sizeBytes <= r.cap {
+			break
+		}
+
+		// dump a message from the queue and try again
+		msg := r.queue[0]
+		r.queue = r.queue[1:]
+		r.sizeBytes -= int64(msg.Size)
```

Comment:
I'd think the simplest thing to do is just drop the message we're trying to enqueue if there's no space.